common/thorhelper/thorstrand.cpp

```diff
@@ -0,0 +1,1353 @@
+        value = NULL;
+        abortSoon = false;
+        done = false;
+    }
+
+    ~OrderedReadAheadQueue()
+    {
+        if (value)
+            value->releaseBlock();
+    }
+
+    void abort()
+    {
+        abortSoon = true;
+    }
+
+    void reset()
+    {
+#ifndef USESPIN
+        space.reinit(1);
+#endif
+        if (value)
+        {
+            value->releaseRows();
+            value->releaseBlock();
+            value = NULL;
+        }
+        abortSoon = false;
+        done = false;
+    }
+
+    bool enqueue(RoxieRowBlock * next)
+    {
+        if (abortSoon)
+            return false;
+#ifdef USESPIN
+        while (hasValue.load(std::memory_order_acquire))
+            _mm_pause();
+#else
+        space.wait();
+#endif
+        if (abortSoon)
+            return false;
+        value = next;
+#ifdef USESPIN
+        hasValue.store(true);
+#else
+        avail.signal();
+#endif
+        return true;
+    }
+
+    void noteWriterStopped()
+    {
+        if (abortSoon)
+            return;
+#ifdef USESPIN
+        while (hasValue.load(std::memory_order_acquire))
+            _mm_pause();
+#else
+        space.wait();
+#endif
+        done = true;
+#ifdef USESPIN
+        hasValue.store(true);
+#else
+        avail.signal();
+#endif
+    }
+
+    bool dequeue(RoxieRowBlock * & ret)
+    {
+        if (abortSoon)
+            return false;
+#ifdef USESPIN
+        while (!hasValue.load(std::memory_order_acquire))
+            _mm_pause();
+#else
+        avail.wait();
+#endif
+        if (abortSoon)
+            return false;
+
+        if (done)
+        {
+#ifndef USESPIN
+            avail.signal();
+#endif
+            return false;
+        }
+
+        ret = value;
+        value = NULL;
+#ifdef USESPIN
+        hasValue.store(false);
+#else
+        space.signal();
+#endif
+        return true;
+    }
+
+protected:
+    RoxieRowBlock * value;
+    bool abortSoon;
+    bool done;
+#ifdef USESPIN
+    std::atomic<bool> hasValue;
+#else
+    Semaphore space __attribute__((aligned(CACHE_LINE_SIZE)));
+    Semaphore avail __attribute__((aligned(CACHE_LINE_SIZE)));
+#endif
+};
+
+
+//I don't think this class really works as it stands there needs to be external control of end of chunk.
+//It needs to share some of the characteristics of the OrderedBranch
+class OrderedReadAheadThread : public CInterface, implements IThreaded
+{
+    //friend class OrderedManyToOneJunction;
+public:
+    OrderedReadAheadThread(RowBlockAllocator & _allocator) : input(NULL), allocator(_allocator)
+    {
+        finished = false;
+        alive = true;
+    }
+
+    virtual void main()
+    {
+        bool done = false;
+        while (!done)
+        {
+            RoxieRowBlock * block = allocator.newBlock();
+            done = block->readFromStream(input);
+            //Purely for testing - this would normally be in an independent callback
+            block->setEndOfChunk();
+            if (block->empty() || !queue.enqueue(block))
```

Comment:
I may have missed it, but where is a RoxieRowBlock exception handled here?
If triggered, empty() will be false, block will be queue and it will continue around this loop - to allocate a new block and read from stream.